domain/application/service/application.go

```diff
@@ -28,14 +30,18 @@ type ApplicationState interface {
 	// satisfying [storageerrors.PoolNotFoundError] if it doesn't exist.
 	GetStoragePoolByName(ctx context.Context, name string) (domainstorage.StoragePoolDetails, error)
 
+	// CreateApplication creates the specified application, with the charm, and
+	// the units if required.
+	CreateApplication(context.Context, string, domaincharm.Charm, ...application.AddUnitArg) (coreapplication.ID, error)
+
 	// UpsertApplication persists the input Application entity.
-	UpsertApplication(context.Context, string, ...application.AddUnitParams) error
+	UpsertApplication(context.Context, string, ...application.AddUnitArg) error
```

Comment:
What are we going to do with this if we have create?

Reply:
No idea, this is up to @wallyworld to decide. I'm just removing the concept of `Upsert` because it doesn't make any sense in terms of applications i.e. you need a charm when creating an application and then you don't touch a charm from then on. Upsert confuses that.